iOS Voice Assistant

ErisMik opened the pull request.

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/Info.plist:
```diff
@@ -0,0 +1,51 @@
+	<key>LSRequiresIPhoneOS</key>
+	<true/>
+	<key>NSMicrophoneUsageDescription</key>
+	<string>For voice activity detection</string>
```
For voice assistant

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/AudioPlayerStream.swift:
```diff
@@ -0,0 +1,78 @@
+import Foundation
```
Since the other source files have our copyright header, you might as well include it here too

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+
+    @Published var enginesLoaded = false
+
+    static let statusTextDefault = "Say `Picovoice` to start!"
```
Since we show this every time before transcription/prompting, it makes more sense to just omit the 'to start'
```suggestion
    static let statusTextDefault = "Say `Picovoice`!"
```

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    public func unloadEngines() {
+        do {
+            try VoiceProcessor.instance.stop()
+            VoiceProcessor.instance.clearFrameListeners()
+            VoiceProcessor.instance.clearErrorListeners()
+        } catch {
+            DispatchQueue.main.async { [self] in
+                errorMessage = "\(error.localizedDescription)"
+            }
+        }
```
Does the voice processor record even when we're generating the completion and doing TTS? If that's the case, I think it would be better to start it on wakeword detect and stop it after cheetah finishes. This way we're running fewer threads and closing the mic stream when we're not using it

ErisMik replied:
Actually, on iOS it seems that stopping and starting voice processor is a fairly expensive call. It seems to interrupt audio playback which makes a poor user experience.

laves replied:
That's odd... it shouldn't be that expensive and it really shouldn't affect the audio playback since stopping the input stream should happen entirely before the output stream starts. 

State should be:
`recording start -> wake word -> stt -> recording stop -> llm generate -> tts -> ...back to beginning`

It kind of sends a weird message to devs looking at this if we keep the mic stream open when it is not using audio. Especially if they use it as a basis for a project where they introduce even more app logic to the llm/tts portion. If there is a way to keep it open only when we need it, that would be preferable.

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+                        chatState = .STT
+                    }
+                }
+            }
+            if chatState == .STT {
```
else if?

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+                if keyword != -1 {
+                    DispatchQueue.main.async { [self] in
+                        statusText = "Listening..."
+                        chatText.append(Message(speaker: "You", msg: ""))
```
```suggestion
                        chatText.append(Message(speaker: "You:", msg: ""))
```

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+
+                let result = try picollm!.generate(
+                    prompt: dialog!.prompt(),
+                    completionTokenLimit: 128,
+                    streamCallback: streamCallback)
```
you can add stop phrases here

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+
+    private func streamCallback(completion: String) {
+        DispatchQueue.main.async { [self] in
+            completionQueue.async {
+                self.completionArray.append(completion)
+            }
+            chatText[chatText.count - 1].append(text: completion)
```
since we have the list of stop phrases, I filter them out so that we don't print or speak them

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift:
```diff
@@ -0,0 +1,335 @@
+                try dialog!.addHumanRequest(content: chatText[chatText.count - 1].msg)
+
+                DispatchQueue.main.async { [self] in
+                    chatText.append(Message(speaker: "picoLLM", msg: ""))
```
```suggestion
                    chatText.append(Message(speaker: "picoLLM:", msg: ""))
```

laves commented on recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/PicoLLMVoiceAssistantDemoApp.swift:
```diff
@@ -0,0 +1,19 @@
+import SwiftUI
+
+@main
+struct PicoLLMVoiceAssistantDemoApp: App {
```
I think these cookbook demos aren't supposed to be as branded and picoLLM focussed (see the readme in the llm-voice-assistant dir 'LLM-Powered Voice Assistant'). I think we should just call this demo LLM Voice Assistant, like the demo directory is named.

laves commented on recipes/llm-voice-assistant/ios/README.md:
```diff
@@ -0,0 +1,51 @@
+
+5. Press the `Load Model` button and load the model file from your device's storage.
+
+6. Say "picovoice", then speak to the voice assistant!
```
```suggestion
6. Say "Picovoice", then speak to the voice assistant!
```

laves commented on res/.lint/spell-check/.cspell.json:
```diff
@@ -11,6 +11,18 @@
-    "**/__pycache__/**"
+    "**/__pycache__/**",
+
+    // ios
```
I think you have to add swift codestyle as an action

ksyeo1010 commented on recipes/llm-voice-assistant/ios/LLMVoiceAssistantDemo/ChatView.swift:
```diff
@@ -0,0 +1,115 @@
+
+        ZStack {
+            VStack(alignment: .center) {
+                Text("picoLLM Voice Assistant Demo")
```
```suggestion
                Text("Voice Assistant Demo")
```

ksyeo1010 commented:
I am getting some `failed to register bundle identifier` errors, does this happened to you?

ErisMik commented:
> I am getting some `failed to register bundle identifier` errors, does this happened to you?

Yeah this happens with all our demos, you have to change the bundle identifier to something like `ai.picovoice.LLMVoiceAssistantDemo.Kwangsoo` or something on your machine.

ErisMik commented:
@laves Can't seem to reply to your comment. Fixed it using the callbacks correctly, we now only record when needed. Also fixed a bug with not flushing the warmupPCM if it never reaches the 1 second mark.
